### DISCUSSION TITLE
Single Sign-On and Seamless SSO?

### DISCUSSION THREAD

**Discussion by chrisbartley** [directus/directus#12055] [Archive]
Apologies if I'm just being incredibly dense, but I'm hoping someone could help shed a little more light on how [Seamless SSO](https://docs.directus.io/configuration/sso/#seamless-sso) is supposed to work?  In particular, it's entirely unclear to me how to tell the Directus SDK that it needs to use the refresh token from the cookie being set by the Directus server after successful SSO.

Here's the high-level description of my current architecture, based very closely on the [Directus SSO instructions](https://docs.directus.io/configuration/sso/#openid):

### Directus Server
- localhost:8055
- Directus v9.6.0
- MySQL v5.7.19
- Node.js v16.14.0
- SSO configuration from my `config-dev.js`:

```javascript
   ACCESS_TOKEN_TTL : "15m",
   REFRESH_TOKEN_TTL : "7d",
   REFRESH_TOKEN_COOKIE_NAME : "directus_refresh_token",

   AUTH_PROVIDERS : "google",

   AUTH_GOOGLE_DRIVER : "openid",
   AUTH_GOOGLE_CLIENT_ID : "************-********************************.apps.googleusercontent.com",
   AUTH_GOOGLE_CLIENT_SECRET : "******-******************-*********",
   AUTH_GOOGLE_ISSUER_URL : "https://accounts.google.com",
   AUTH_GOOGLE_IDENTIFIER_KEY : "email",
   AUTH_GOOGLE_ICON : "google",
   AUTH_GOOGLE_ALLOW_PUBLIC_REGISTRATION : true, 
   AUTH_GOOGLE_DEFAULT_ROLE_ID : "********-****-****-****-************", 
   AUTH_GOOGLE_PARAMS : {hd: '*************'},

   REFRESH_TOKEN_COOKIE_DOMAIN : "localhost",
   REFRESH_TOKEN_COOKIE_SECURE : true,
   REFRESH_TOKEN_COOKIE_SAME_SITE : "None",
```

### Client Web App
- localhost:8989
- `@directus/sdk` v9.6.0
- Plain old Node.js/Express web app
- Node.js v16.14.0
- Handler for the `/login` route:

```JavaScript
const express = require('express');
const router = express.Router();
const { Directus } = require('@directus/sdk');

module.exports = function() {
   router.get('/', function(req, res, next) {
      Promise.resolve()
            .then(async function() {
               const directus = new Directus('http://localhost:8055');

               // I can't seem to figure out how to make this work...see discussion below
               // await directus.auth.refresh();

               console.log("====================================================");
               console.log('directus.auth.token: [' + directus.auth.token + ']');
               console.log(directus.auth);
               console.log('Cookie: ' + req.cookies['directus_refresh_token']);

               res.render('login/index', { title : "Login", authToken : directus.auth.token });
            })
            .catch(e => {
               console.log(e);
               res.render('login/index', { title : "Login FAILED" });
            })
   });

   return router;
};
```

The SSO part of all this works great.  I have a link in the client web app like this:

```html
<a href="http://localhost:8055/auth/login/google?redirect=http://localhost:8989/login">Login</a>
```

Clicking that link takes me to my SSO provider's login screen, as expected.  I enter my credentials (a Google Workspace account), it does the little redirection dance, and then eventually lands on `http://localhost:8989/login`.  And inspecting the cookies, I do indeed see the cookie named `directus_refresh_token`. And the client web app sees it too, as shown by the output from the `/login` route in the terminal window:

```
directus.auth.token: [null]
Auth {
  mode: 'json',
  autoRefresh: true,
  msRefreshBeforeExpires: 30000,
  staticToken: '',
  timer: null,
  _transport: Transport {
    config: {
      url: 'http://localhost:8055',
      beforeRequest: [Function: beforeRequest]
    },
    axios: [Function: wrap] {
      request: [Function: wrap],
      getUri: [Function: wrap],
      delete: [Function: wrap],
      get: [Function: wrap],
      head: [Function: wrap],
      options: [Function: wrap],
      post: [Function: wrap],
      put: [Function: wrap],
      patch: [Function: wrap],
      defaults: [Object],
      interceptors: [Object],
      create: [Function: create]
    },
    beforeRequest: [Function: beforeRequest]
  },
  _storage: MemoryStorage { prefix: '', values: {} }
}
Cookie: iOTllJzAF5lHRwPJ4JZa-vQOXKd41XYnUEiNHuAI03FBx5oTY6oHXfWAv6-ofc_-
GET /login 200 - 42.155 ms -
```

Even more, if I open a tab and go to the Directus server (localhost:8055), it knows the user is logged in, and lets me poke around (I have `AUTH_GOOGLE_ALLOW_PUBLIC_REGISTRATION` set to `true`, and the default role has app access enabled and read-only privileges).  So I'm definitely logged in, and all that is working just fine.

What's _not_ working is the "seamless" part, where the client web app should (I think?) be able to use the Directus SDK to make authenticated requests to the Directus server.  Step 3 of the [Seamless SSO docs](https://docs.directus.io/configuration/sso/#seamless-sso) make it sound to me like all I have to do is have the `/login` route create an instance of the Directus SDK, giving it the URL of the Directus server, and then call `await directus.auth.refresh();`. But that blows up because it's getting an HTTP 400 back from the Directus server, with this error:

```
TransportError: "refresh_token" is required in either the JSON payload or Cookie
    at Transport.request (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/@directus/sdk/dist/sdk.cjs.js:614:23)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Transport.post (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/@directus/sdk/dist/sdk.cjs.js:643:16)
    at async Auth.refresh (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/@directus/sdk/dist/sdk.cjs.js:725:26)
    at async /Users/chris/Documents/Work/Projects/phys-comp-maker-cards/routes/login.js:11:16 {
  parent: Error: Request failed with status code 400
      at createError (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/axios/lib/core/createError.js:16:15)
      at settle (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/axios/lib/core/settle.js:17:12)
      at IncomingMessage.handleStreamEnd (/Users/chris/Documents/Work/Projects/phys-comp-maker-cards/node_modules/axios/lib/adapters/http.js:293:11)
      at IncomingMessage.emit (node:events:532:35)
      at endReadableNT (node:internal/streams/readable:1346:12)
      at processTicksAndRejections (node:internal/process/task_queues:83:21) {
    config: {
      transitional: [Object],
      adapter: [Function: httpAdapter],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      maxBodyLength: -1,
      validateStatus: [Function: validateStatus],
      headers: [Object],
      baseURL: 'http://localhost:8055',
      withCredentials: true,
      method: 'post',
      url: '/auth/refresh',
      data: '{"refresh_token":null}'
    }
...
```

So, yeah, that makes total sense that the Directus SDK instance I created wouldn't know the refresh token's value--and I see from the config there in the stack trace that it's null--but it's entirely unclear to me how I'm supposed to tell it what the refresh token is before calling `directus.auth.refresh()`.

Am I missing something totally obvious???  Thanks so much for your help (and patience reading through all this)!

**Comment by chrisbartley** [comment]
Update!  OK, after some digging, I at least found a way to give the refresh token to the Directus SDK before making the call to `directus.auth.refresh()`.  I'm just stuffing it into the storage object, like this:

```JavaScript
directus.storage.auth_refresh_token = req.cookies['directus_refresh_token'];
```

That works, but is it the correct and recommended thing to do?  If so I'm happy to create a PR so this is documented in the Seamless SSO section of the docs.

Here's the working login handler:

```JavaScript
const express = require('express');
const router = express.Router();
const { Directus } = require('@directus/sdk');
const config = require("../config");

const API_ROOT_URL = config.get('cms:apiRootUrl');

module.exports = function() {
   router.get('/', function(req, res, next) {
      Promise.resolve()
            .then(async function() {
               const directus = new Directus(API_ROOT_URL);

               // put the refresh token into storage so that it gets passed to the Directus server when we call refresh
               directus.storage.auth_refresh_token = req.cookies['directus_refresh_token'];

               // call refresh in order to get the auth token...but remember that doing so yields a new refresh token
               await directus.auth.refresh();

               console.log('Refresh token (OLD, from cookie) : ' + req.cookies['directus_refresh_token']);
               console.log('Refresh token (NEW, from refresh): ' + directus.storage.auth_refresh_token);
               console.log('Auth token: ' + directus.auth.token);

               // Overwrite the cookie containing the refresh token, so that the cookie has the correct one
               res.cookie('directus_refresh_token',
                          directus.storage.auth_refresh_token,
                          {
                             domain : 'localhost',                // TODO: get this from config, make sure it's identical to the domain that the Directus server is using
                             maxAge : 7 * 24 * 60 * 60 * 1000,    // 7 days, in millis
                             httpOnly : true,
                             secure : true,
                             sameSite : "None"
                          });

               res.render('login/index', { title : "Login", authToken : directus.auth.token });
            })
            .catch(e => {
               console.log(e);
               res.render('login/index', { title : "Login FAILED" });
            })
   });

   return router;
};
```

You'll see that I call `res.cookie()` after obtaining the auth token.  That's just so that the cookie has the correct refresh token, which is critical if users expect to be able to, say, 1) hit reload while on the `/login` screen; or 2) open another tab and go to the Directus server and have it know that they're logged in.

But all this brings up another question:  What's the best practice, or at least typical practice, after all this?  Do people employing similar, two-site architectures with SSO simply do the `directus.auth.refresh()` dance (getting the refresh token from the cookie, and then calling refresh() to obtain the auth token) for any request which needs to call a method in the SDK which requires auth?  Or do people enable sessions on the client site, and keep auth token and refresh token in the session?  The former is lighter weight for the client app, but definitely more chatty with the Directus server what with all the token refreshing going on.

Thanks!

**Comment by chrisbartley** [comment]
Continuing this conversation with myself, I'm now wondering whether the examples in the docs are for Directus SDK usage on the _front end_ (i.e. browser) of the client site, and not the back end (Node.js), and that's why I couldn't get it to work without manually stuffing the refresh token into the storage object?  That is, I could imagine that an instance of the Directus SDK running in the browser could just magically read/edit the refresh token cookie using standard browser cookie APIs, whereas when used in Node.js it needs to be explicitly given the refresh token since there's no obvious way for it to know how to pick it out of the cookie in the request.

Still, I'd really love it if someone could confirm that, when using the SDK in Node.js, manually inserting it (`directus.storage.auth_refresh_token = req.cookies['directus_refresh_token'];`) is the best method.

**Reply by Thermal-Shock** [reply]
I'll second this request for clarification. We currently use openid and are currently evaluating Directus, so its important to know the definitive way of doing this so everything is secure and we have piece of mind. All the best.

**Reply by chrisbartley** [reply]
FWIW, I've been extremely happy with Directus. I evaluated at least a dozen other options and, for my use case at least, Directus was far and away the best choice. And I've been really happy with the active and extremely responsive developer community (ok, well...this thread isn't exactly a shining example of that).

Anyway, although my questions above remain to be answered definitively, SSO is at least working nicely for me now with my two-site architecture (they share domain name, differing only in subdomain). The most important bits in getting it to work are above, but I'm happy to answer questions or share tips if you like.

**Reply by rijkvanzanten** [reply]
> (ok, well...this thread isn't exactly a shining example of that).

haha mostly because it takes me a minute to really understand the question and come up with a reproduction 😬

**Reply by chrisbartley** [reply]
> haha mostly because it takes me a minute to really understand the question and come up with a reproduction 😬

Sorry, yes, I was just joking around 😂  I realize the fault is all mine, since I buried the questions within a ton of information.  But my questions basically boil down to the following:

- It appears to me that the examples in the Seamless SSO docs may be assuming an SPA architecture, with the Directus SDK instance living in the browser and thus only rarely recreated, yes?  If so, I'm happy to submit a PR to clarify this, and include other examples (for non-SPA), if that's helpful/desired.
- I'm not using an SPA architecture, so my Directus SDK instance lives server side, in the client web app. And so instantiating it and doing the token refresh dance with every request (well, at least every request which needs auth for communicating with the Directus server) is definitely non-optimal. I'll likely end up just enabling sessions and storing auth and refresh token in the session and update/invalidate them as needed. But I'm curious whether others have better/different solutions?
- Because my Directus SDK instances live server side, I'm looking for guidance on how best to tell it what the refresh token is for times when I do need to refresh to get the auth token. My [post above](https://github.com/directus/directus/discussions/12055#discussioncomment-2327698) describes how I'm doing it by setting it in the storage object, but I'm not super confident that's the best way.

Thanks!

**Reply by aidenfoxx** [reply]
@chrisbartley You are correct in that the Seamless SSO section is intended more for SPAs. It seems to rely on the fact that the users refresh cookie will always be sent with SDK "refresh" requests by virtue of being in the the browser. I guess the hop to your SDK server is causing the cookie header to be dropped on subsequent SDK requests to Directus?

If that is the case, you could solve this by writing a custom "Transport" which maps the incoming cookie header to the SDK requests. See https://docs.directus.io/reference/sdk/#transport and https://github.com/directus/directus/blob/main/packages/sdk/src/base/transport.ts#L47?

I might be waaaaaay off here, so please tell me if I'm being stupid @rijkvanzanten.

**Reply by chrisbartley** [reply]
> If that is the case, you could solve this by writing a custom "Transport"

Ah, interesting, thanks!

> I guess the hop to your SDK server is causing the cookie header to be dropped on subsequent SDK requests to Directus?

Well, to be precise, SDK requests to Directus don't require cookies. Not directly, at least.  For example, the call to `directus.auth.refresh()` passes the [refresh token as message body](https://docs.directus.io/reference/authentication/#refresh). So, yeah, the refresh token is obtained from the cookie, but my problem stems from the Directus SDK instance not knowing how to read the cookie, rather than the cookie not being present.

But anyway, with my setup, the cookie is being set and passed around between Directus site and client site just fine.  It's working with localhost (Directus at localhost:8055 and client site at localhost:8989), but most importantly it's working in production (e.g. directus.myserver.com and site.myserver.com), because I have `REFRESH_TOKEN_COOKIE_DOMAIN` set to `myserver.com`.  So that part is all good. 👍  

The primary issue is that, since my client site isn't an SPA and I am thus creating the Directus SDK instance (`const directus = new Directus('directus.myserver.com');`) on the server side of the client site (a new instance with each request, of course), that instance has no way of magically knowing how to read cookies from the request.  I'm using Express.js, and so picking the refresh token from the request is simply `req.cookies['directus_refresh_token']`.  But, well, then what?  What's the correct way to give it to the SDK instance to ensure it gets passed to the Directus server when making requests such as `directus.auth.refresh()`?  I found that this works:

```javascript
directus.storage.auth_refresh_token = req.cookies['directus_refresh_token'];
```
But I'm unsure whether that's merely a "wellllll, it works, but it's gross" hack.  So I'm looking for guidance on best practice here, perhaps doing it differently with a custom Transport like you suggested.

Thanks!

**Reply by aidenfoxx** [reply]
> Well, to be precise, SDK requests to Directus don't require cookies. Not directly, at least. For example, the call to directus.auth.refresh() passes the refresh token as message body

That is true, but also not required. Directus will take the "refresh_token" from the request body, or fallback to the refresh token cookie. See https://github.com/directus/directus/blob/main/api/src/controllers/auth.ts#L70

So by supplying the cookie header via the transport to the SDK, the refresh call should work as expected with less funny business. Obviously not an ideal solution, but it feels a little cleaner than manually setting values into storage. 😄

**Reply by aidenfoxx** [reply]
Looking more deeply into the transport, it's not actually easy to extend the default transport. Seems like it's intended mostly to write different transport layers. I also see an additional issue that you must respond with the new "refresh_token" cookie. Really doesn't feel like the SDK was built with this in mind... 😅

**Reply by chrisbartley** [reply]
> That is true, but also not required. Directus will take the "refresh_token" from the request body, or fallback to the refresh token cookie. See https://github.com/directus/directus/blob/main/api/src/controllers/auth.ts#L70

Oh wow 😮  OK, so a Directus SDK instance living in the front end (browser) really can be brainless, with no need to worry about reading the cookie or even sending a message body. Neat, I guess.

> I also see an additional issue that you must respond with the new "refresh_token" cookie.

Yep.  I have that problem solved, at least, by manually setting it with the new value after calling `directus.auth.refresh()`.

Thanks again for your help! I'll poke around in Transport and such some more...

**Reply by rijkvanzanten** [reply]
@azrikahar is also currently looking into tweaking the SDK to make the auto-refreshing of the tokens a little more stable, so any and all suggestions you have into optimizing the transports setup is highly appreciated 🙂

**Reply by aidenfoxx** [reply]
@rijkvanzanten it would be really helpful to be able to extend the default Transport IMO. You might not want to re-invent the wheel just to add a few headers to your requests or something. 😄

**Reply by rijkvanzanten** [reply]
Yeah absolutely. I also think we should just expose the underlying axios instance and don't be too silly with these transport object classes. KISS

**Comment by keesvanbemmel** [comment]
@chrisbartley I think I'm running into the same issues you had. Been trying to get seamless SSO to work with this Nuxt3 starter project: https://github.com/bryantgillespie/nuxt3-directus-starter

Can't seem to find a way to get the SDK to refresh, it's giving me the same error `"refresh_token" is required in either the JSON payload or Cookie` you had.

Any additional pointers? This is a SSR app, so it's not only the client that connects to/with Directus.

**Reply by rijkvanzanten** [reply]
@bryantgillespie?!

**Reply by keesvanbemmel** [reply]
It looks like the SDK is looking for a "auth_refresh_token" cookie, but I only have a "directus_refresh_token"

**Comment by alsherif-khalaf** [comment]
I tried to us Nextjs middleware :

```ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export async function middleware(request: NextRequest) {
  const response = NextResponse.next();
  const userURL = 'https://directus.com/cms/users/me';
  const refreshURL = 'https://directus.com/cms/auth/refresh';

  console.log("request", request.nextUrl.pathname)

  try {
    let accessToken = request.cookies.get('access_token')?.value;
    const refreshToken = request.cookies.get('auth_refresh_token')?.value;

    if (!accessToken && !refreshToken) {
      response.cookies.set('auth', 'false');
      return response; // User is not logged in, stop execution
    }

    const currentTimestamp = Date.now();
    const accessExpires = 360000; // Expiration time for access token in milliseconds
    const refreshExpires = 86400000; // Expiration time for refresh token in milliseconds
    const accessExpirationTimestamp = currentTimestamp + accessExpires;
    const refreshExpirationTimestamp = currentTimestamp + refreshExpires;

    if (request.method === 'GET' || request.method === 'POST') {
      if (accessToken && currentTimestamp < accessExpirationTimestamp) {
        // Access token is still valid, proceed with user authentication
        console.log("test accessToken")
        // check if access token valid
        const userResponse = await fetch(userURL, {
          headers: {
            Authorization: `Bearer ${accessToken}`
          }
        });

        if (userResponse.ok) {
          // User data successfully fetched and authenticated
          const userData = await userResponse.json()
          response.cookies.set('auth', 'true');
          response.cookies.set('id', userData.data.id);
          response.cookies.set('role', userData.data.role);
        }
      } else if (refreshToken && currentTimestamp < refreshExpirationTimestamp) {
        const refreshResponse = await fetch(refreshURL, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            refresh_token: refreshToken,
            mode: 'json'
          })
        });

        if (refreshResponse.ok) {
          const tokenData = await refreshResponse.json();
          console.log("tokenData", tokenData);
          const newAccessToken = tokenData.data.access_token;
          const newRefreshToken = tokenData.data.refresh_token;
          const newExpires = tokenData.data.expires;
          const newAccessExpirationTimestamp = currentTimestamp + newExpires;
          const newRefreshExpirationTimestamp = currentTimestamp + refreshExpires;

          response.cookies.set('auth_refresh_token', newRefreshToken, { maxAge: 86400 }); // Expires after 1 day
          response.cookies.set('access_token', newAccessToken, { expires: newAccessExpirationTimestamp });
          response.cookies.set('auth', 'true');

          // Check if the new access token is valid by making a test request
          const userResponse = await fetch(userURL, {
            headers: {
              Authorization: `Bearer ${newAccessToken}`
            }
          });

          if (userResponse.ok) {
            // New access token is valid
            const userData = await userResponse.json()
            response.cookies.set('auth', 'true');
            response.cookies.set('id', userData.data.id);
            response.cookies.set('role', userData.data.role);
            return response;
          }
        }
      }
    }
  } catch (error) {
    console.error('Token refresh failed:', error);
  }

  return response;
}

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!api|_next/static|_next/image|favicon.ico|logo.png|hero.webp).*)',
  ],
};
  
```

i changed conf in .env fie :

```
# The duration that the access token is valid ["15m"]
ACCESS_TOKEN_TTL="6m"

# The duration that the refresh token is valid, and also how long users stay logged-in to the App ["7d"]
REFRESH_TOKEN_TTL="1d"
```

I barely understand what happens ..
I will test this live
